### PR TITLE
Add an actual quantity concept and use it in class Multivector

### PR DESCRIPTION
**Pull request by eggrobin**
We could use it more even in `grassman.hpp`, on the functions; but I am sending this first cut first to see how we like it before I change things on a larger scale.

With this, the error message on
```C++
  Vector<World, Acceleration> γ;  // Error: Should be Vector<Acceleration, World>.
```
changes as follows; see also https://devblogs.microsoft.com/cppblog/structured-diagnostics-in-the-new-problem-details-window/.

## Before

(`Norm²` fails to compile.)

```
1>grassmann_test.cpp
1>C:\Users\robin\Projects\mockingbirdnest\Principia\quantities\named_quantities.hpp(27,65): error C2794: 'Type': is not a member of any direct or indirect base class of 'principia::quantities::_generators::internal::ExponentiationGenerator<Scalar,2>'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\quantities\named_quantities.hpp(27,65): error C2794:         with
1>C:\Users\robin\Projects\mockingbirdnest\Principia\quantities\named_quantities.hpp(27,65): error C2794:         [
1>C:\Users\robin\Projects\mockingbirdnest\Principia\quantities\named_quantities.hpp(27,65): error C2794:             Scalar=principia::geometry::GrassmannTest::World
1>C:\Users\robin\Projects\mockingbirdnest\Principia\quantities\named_quantities.hpp(27,65): error C2794:         ]
1>(compiling source file 'grassmann_test.cpp')
1>C:\Users\robin\Projects\mockingbirdnest\Principia\quantities\named_quantities.hpp(27,65):
1>the template instantiation context (the oldest one first) is
1>	C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(72,31):
1>	see reference to class template instantiation 'principia::geometry::_grassmann::internal::Multivector<principia::geometry::GrassmannTest::World,principia::quantities::_named_quantities::internal::Acceleration,1>' being compiled
1>	C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann.hpp(55,3):
1>	see reference to alias template instantiation 'principia::quantities::_named_quantities::internal::Square<Scalar>' being compiled
1>        with
1>        [
1>            Scalar=principia::geometry::GrassmannTest::World
1>        ]
1>	C:\Users\robin\Projects\mockingbirdnest\Principia\quantities\named_quantities.hpp(29,16):
1>	see reference to alias template instantiation 'principia::quantities::_named_quantities::internal::Exponentiation<Scalar,2>' being compiled
1>        with
1>        [
1>            Scalar=principia::geometry::GrassmannTest::World
1>        ]
1>C:\Users\robin\Projects\mockingbirdnest\Principia\quantities\named_quantities.hpp(29,16): error C2938: 'principia::quantities::_named_quantities::internal::Exponentiation' : Failed to specialize alias template
1>(compiling source file 'grassmann_test.cpp')
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1): error C2282: 'principia::geometry::GrassmannTest::~GrassmannTest' cannot override 'testing::Test::~Test'
1>C:\Users\robin\Projects\mockingbirdnest\Google\googletest\googletest\include\gtest\gtest.h(419,12):
1>'testing::Test::~Test' is not deleted
```

<details>
<summary>More weird errors about destructors…</summary>

```
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(85,1): error C2282: 'principia::geometry::GrassmannTest_VectorFMA_Test::~GrassmannTest_VectorFMA_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Google\googletest\googletest\include\gtest\internal\gtest-internal.h(562,1): error C2282: 'testing::internal::SuiteApiResolver<principia::geometry::GrassmannTest>::~SuiteApiResolver' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>(compiling source file 'grassmann_test.cpp')
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Google\googletest\googletest\include\gtest\internal\gtest-internal.h(562,1):
1>the template instantiation context (the oldest one first) is
1>	C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(85,1):
1>	see reference to class template instantiation 'testing::internal::SuiteApiResolver<principia::geometry::GrassmannTest>' being compiled
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(104,1): error C2282: 'principia::geometry::GrassmannTest_BivectorFMA_Test::~GrassmannTest_BivectorFMA_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(123,1): error C2282: 'principia::geometry::GrassmannTest_TrivectorFMA_Test::~GrassmannTest_TrivectorFMA_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(142,1): error C2282: 'principia::geometry::GrassmannTest_Operators_Test::~GrassmannTest_Operators_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(151,1): error C2282: 'principia::geometry::GrassmannTest_SpecialOrthogonalLieAlgebra_Test::~GrassmannTest_SpecialOrthogonalLieAlgebra_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(160,1): error C2282: 'principia::geometry::GrassmannTest_MixedScalarMultiplication_Test::~GrassmannTest_MixedScalarMultiplication_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(178,1): error C2282: 'principia::geometry::GrassmannTest_VectorSpaces_Test::~GrassmannTest_VectorSpaces_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(223,1): error C2282: 'principia::geometry::GrassmannTest_GrassmannAlgebra_Test::~GrassmannTest_GrassmannAlgebra_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(251,1): error C2282: 'principia::geometry::GrassmannTest_Actions_Test::~GrassmannTest_Actions_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(268,1): error C2282: 'principia::geometry::GrassmannTest_Norm_Test::~GrassmannTest_Norm_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(279,1): error C2282: 'principia::geometry::GrassmannTest_Normalize_Test::~GrassmannTest_Normalize_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(304,1): error C2282: 'principia::geometry::GrassmannDeathTest_SerializationError_Test::~GrassmannDeathTest_SerializationError_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(329,1): error C2282: 'principia::geometry::GrassmannTest_SerializationSuccess_Test::~GrassmannTest_SerializationSuccess_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(361,1): error C2282: 'principia::geometry::GrassmannTest_Angles_Test::~GrassmannTest_Angles_Test' cannot override 'principia::geometry::GrassmannTest::~GrassmannTest'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(81,1):
1>'principia::geometry::GrassmannTest::~GrassmannTest' is deleted
1>Done building project "geometry.vcxproj" -- FAILED.
```

</details>

## After

```
1>grassmann_test.cpp
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann_test.cpp(72,3): error C7602: 'principia::geometry::_grassmann::internal::Vector': the associated constraints are not satisfied
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann.hpp(141,1):
1>see declaration of 'principia::geometry::_grassmann::internal::Vector'
1>C:\Users\robin\Projects\mockingbirdnest\Principia\geometry\grassmann.hpp(140,10):
1>the concept 'principia::quantities::_concepts::internal::quantity<principia::geometry::GrassmannTest::World>' evaluated to false
1>	C:\Users\robin\Projects\mockingbirdnest\Principia\quantities\concepts.hpp(53,20):
1>	the concept 'principia::base::_traits::instance<principia::geometry::GrassmannTest::World,principia::quantities::_quantities::internal::Quantity>' evaluated to false
1>		C:\Users\robin\Projects\mockingbirdnest\Principia\base\traits.hpp(71,20):
1>		the constraint was not satisfied
1>	C:\Users\robin\Projects\mockingbirdnest\Principia\quantities\concepts.hpp(53,45):
1>	the concept 'std::same_as<principia::geometry::GrassmannTest::World,double>' evaluated to false
1>		C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.38.33130\include\concepts(35,9):
1>		'principia::geometry::GrassmannTest::World' and 'double' are different types
1>Done building project "geometry.vcxproj" -- FAILED.
```

### CODE REVIEW

File: base/traits.hpp [hunk 1]
```diff
@@ -59,6 +59,11 @@ using internal::all_different_v;
 template<template<typename...> typename T, typename U>
 inline constexpr bool is_instance_of_v = internal::is_instance_of<T, U>::value;
 
+// Note the argument inversion so that we can use instance_of<T> in a requires
```
Comment (pleroy) [inline base/traits.hpp:62]: I don't quite understand this inversion.  I would argue that:
```
requires instance_of<Quantity, T>
```
would read better than:
```
requires instance_of<T, Quantity>
```
If there is a technical reason for the inversion, it should be explained here (I couldn't find one, see https://godbolt.org/z/fYbndKd1c) and we should also invert the `is_instance_of` trait to minimize confusion.

**Review by pleroy — COMMENTED** [review]
You'll need to remove ksp_plugin_adapter_stub (again!).

File: base/traits.hpp (hunk 1 above)
Comment (eggrobin) [reply base/traits.hpp:62]: In `requires` clauses, you want to be able to write
`{ expression } -> instance_of<Quantity>`

Compare the inversion between https://en.cppreference.com/w/cpp/types/is_base_of and https://en.cppreference.com/w/cpp/concepts/derived_from.

File: base/traits.hpp [hunk 2]
```diff
@@ -59,6 +59,17 @@ using internal::all_different_v;
 template<template<typename...> typename T, typename U>
 inline constexpr bool is_instance_of_v = internal::is_instance_of<T, U>::value;
 
+// Note that the order of template parameters is backward from is_instance_of_v.
+// This makes it possible to write
+//   requires { { expression } -> instance<T>; }
+// or
+//   template<instance<T> U>
+// but is_instance_of_v<T, U> should be preferred in boolean expressions.
+// An exception is when defining concepts; there we use instance_of<U, T> so as
```
Comment (pleroy) [inline base/traits.hpp:68]: `instance<U, T>`

**Pull request merged**